### PR TITLE
Working BooleanLineVectorizer

### PR DESCRIPTION
**Pull request by WoutLegiest**
- Implementation of the BoolLineVectorizer, a pass in the CGGI dialect
- Adding a test for the pass above
- Expansion of the CGGI Dialect to support the representation of gates through attributes
- Small refactoring of the TfheRustBool Dialect

### CODE REVIEW

File: lib/Dialect/CGGI/IR/CGGIOps.td [hunk 1]
```diff
@@ -57,6 +59,17 @@ def CGGI_NotOp : CGGI_Op<"not", [
   let summary = "Logical NOT of two ciphertexts";
 }
 
+def CGGI_PackedOp : CGGI_Op<"packed", [
+    Pure]> {
```
Comment (asraa) [inline lib/Dialect/CGGI/IR/CGGIOps.td:63]: This is also a Commutative operation, right? I also think you could use `SameOperandsAndResultType` - even if you have an attribute, i think that the type matching only occurs on values (I found an example, at least: https://github.com/google/jax/blob/b1066ee41373991bb66c5a56b2874dc5455b7f70/jaxlib/mosaic/dialect/tpu/tpu.td#L237)

File: lib/Dialect/CGGI/IR/CGGIOps.td [hunk 2]
```diff
@@ -57,6 +59,17 @@ def CGGI_NotOp : CGGI_Op<"not", [
   let summary = "Logical NOT of two ciphertexts";
 }
 
+def CGGI_PackedOp : CGGI_Op<"packed", [
```
Comment (asraa) [inline lib/Dialect/CGGI/IR/CGGIOps.td:62]: nit: maybe `packed_gates`?

File: lib/Dialect/CGGI/IR/CGGIOps.td [hunk 3]
```diff
@@ -57,6 +59,17 @@ def CGGI_NotOp : CGGI_Op<"not", [
...
+    Pure]> {
+  let arguments = (ins
+    CGGIGate_Attr:$gates,
```
Comment (asraa) [inline lib/Dialect/CGGI/IR/CGGIOps.td:65]: could this op use a verifier to validate that the gate attr is the same length as the lhs/rhs/output? (I think lhs/rhs/output sizes will be verified if you add SameOperandsAndResultType)

File: lib/Dialect/CGGI/IR/CGGIAttributes.td [hunk 4]
```diff
@@ -22,4 +30,20 @@ def CGGI_CGGIParams : AttrDef<CGGI_Dialect, "CGGIParams"> {
   let assemblyFormat = "`<` struct(params) `>`";
 }
 
+def CGGIGate_Attr : CGGI_Attr<"CGGIGate", "cggi_gate"> {
+  let summary = "An Attribute containing an array of strings to store bool gates";
+
+  let description = [{
+    Attribute to store what kind of Boolean gate we are working on
```
Comment (asraa) [inline lib/Dialect/CGGI/IR/CGGIAttributes.td:37]:
```suggestion
    This attributes stores a list of string identifiers for Boolean gates.
    
     This used in the `cggi.packed` operation to indicate the boolean gate that applies pairwise between elements of two ciphertext arrays. For example,
     
     %0 = cggi.packed %a, %b {gates = #cggi.cggi_gate<"and", "xor">} : tensor<2x!lwe.lwe_ciphertext>

     applies an "and" gate between the first elements of %a and %b and an xor gate between the second elements.
```

File: lib/Dialect/CGGI/Transforms/Passes.td [hunk 5]
```diff
@@ -19,4 +19,20 @@ def SetDefaultParameters : Pass<"cggi-set-default-parameters"> {
   let dependentDialects = ["mlir::heir::cggi::CGGIDialect"];
 }
 
+def BooleanLineVectorizer : Pass<"cggi-boolean-line-vectorize"> {
+  let summary = "Group different logic gates with the packed API";
+  let description = [{
+    This pass groups indepentend logic gates into a single call of the packed
+    operations.
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/Passes.td:26]:
```suggestion
    This pass groups independent logic gates into a single call of the packed
    operations.
```

File: lib/Dialect/CGGI/Transforms/Passes.td [hunk 6]
```diff
@@ -19,4 +19,20 @@ def SetDefaultParameters : Pass<"cggi-set-default-parameters"> {
...
+
+    Pass is intended for the `FPT` tfhe-rs API.
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/Passes.td:28]: nit: might benefit from a short example, or some description that this is an improvement over straight line vectorizer.

File: lib/Dialect/CGGI/Transforms/Passes.td [hunk 7]
```diff
@@ -19,4 +19,20 @@ def SetDefaultParameters : Pass<"cggi-set-default-parameters"> {
...
+  }];
+
+  let dependentDialects = [
+    "tensor::TensorDialect",
+  ];
+
+  let dependentDialects = ["mlir::heir::cggi::CGGIDialect"];
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/Passes.td:35]: I think you can get away without cggi as a dependent dialect since it must have already been loaded as a dialect to parse the input. (https://mlir.llvm.org/getting_started/Faq/#registered-loaded-dependent-whats-up-with-dialects-management)

File: lib/Dialect/TfheRustBool/IR/TfheRustBoolTypes.td [hunk 8]
```diff
@@ -28,4 +28,27 @@ def TfheRustBool_ServerKey : TfheRustBool_Type<"ServerKey", "server_key", [PassB
   let summary = "The boolean server key required to perform homomorphic operations.";
 }
 
+
+class TfheRustBool_BooleanGate<string gate>
+  : TfheRustBool_Type<"BooleanGate" # gate, "bin_gate" # gate> {
+  let summary = "Enum placeholder to define the binary gate" # gate # " to perform as homomoprhic operation.";
+}
+
+foreach i = ["And", "Or", "Xor", "Nand", "Nor", "Xnor"] in {
```
Comment (asraa) [inline lib/Dialect/TfheRustBool/IR/TfheRustBoolTypes.td:37]: this is awesome! (any reason not do something similar for the CGGI attributes? is it more difficult to do with a string attr enum?)

File: lib/Dialect/TfheRustBool/IR/TfheRustBoolOps.td [hunk 9]
```diff
@@ -71,4 +73,19 @@ def MuxOp : TfheRustBool_Op<"mux", [
   let results = (outs TfheRustBool_Encrypted:$output);
 }
 
+
+// ToDo: Change to an only tensor type?
```
Comment (asraa) [inline lib/Dialect/TfheRustBool/IR/TfheRustBoolOps.td:77]: just to make sure, did you want to merge the TfheRustBool files with this PR or the follow-up #842?

File: tests/cggi/boolean_line_vectorizer.mlir [hunk 10]
```diff
@@ -0,0 +1,75 @@
+// RUN: heir-opt --cggi-boolean-line-vectorize %s
+
+#encoding = #lwe.unspecified_bit_field_encoding<cleartext_bitwidth = 1>
+!ct_ty = !lwe.lwe_ciphertext<encoding = #encoding>
+!pt_ty = !lwe.lwe_plaintext<encoding = #encoding>
+
+// CHECK-LABEL: add_one
+// CHECK-COUNT-1: cggi.packed %[[*]], %[[*]] {gates = #cggi.cggi_gate<"and", "xor">} : (tensor<2x!lwe.lwe_ciphertext
```
Comment (asraa) [inline tests/cggi/boolean_line_vectorizer.mlir:8]: oh no! only one cggi.packed haha.

could you also add a syntax test in tests/cggi/ that tests the syntax of this operation so we have it as documentation? (if you add the verifier would also be good to test that error condition on the gate length)

File: lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+namespace mlir {
+namespace heir {
+namespace cggi {
+
+#define GEN_PASS_DEF_BOOLEANLINEVECTORIZER
+#include "lib/Dialect/CGGI/Transforms/Passes.h.inc"
+
+bool areCompatibleBool(Operation *lhs, Operation *rhs) {
+  if (
+      // lhs->getName() != rhs->getName() ||
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp:32]: nit: remove this! (yay!)

File: lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+      lhs->getDialect() != rhs->getDialect() ||
+      lhs->getResultTypes() != rhs->getResultTypes() ||
+      lhs->getAttrs() != rhs->getAttrs()) {
+    return false;
+  }
+  return OpTrait::hasElementwiseMappableTraits(lhs);
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp:38]: this might be somewhat fragile right? since there might be other elementwise operations that aren't boolean gates..? maybe add a comment or a TODO if this can be made better with a BooleanPackableGate trait on the CGGI_BinaryGateOp's?

File: lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+          str = "nand";
+        } else if (isa<cggi::XorOp>(op)) {
+          str = "xor";
+        } else if (isa<cggi::XNorOp>(op)) {
+          str = "xnor";
+        } else if (isa<cggi::OrOp>(op)) {
+          str = "or";
+        } else if (isa<cggi::NorOp>(op)) {
+          str = "nor";
+        }
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp:138]: might be worth a final if statement that fails / prints LLVM_DEBUG statement in case there is an incompatible operation that we haven't added yet?

File: lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+        llvm::dbgs() << "\n";
+      });
+
+      auto oplist = CGGIGateAttr::get(&context, vectorizedGateOperands);
+
+      auto vectorizedOp = builder.create<cggi::PackedOp>(
+          key->getLoc(), tensorType, oplist, vectorizedOperands[0],
+          vectorizedOperands[1]);
+
+      vectorizedOp->getResult(0).setType(tensorType);
```
Comment (asraa) [inline lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp:183]: interesting! just curious, why was this needed? I would assume the first operand in the builder above was for the result type.

**Review by asraa — COMMENTED** [review]
Nice job!!!

File: lib/Dialect/TfheRustBool/IR/TfheRustBoolOps.td (hunk 9 above)
Comment (WoutLegiest) [reply lib/Dialect/TfheRustBool/IR/TfheRustBoolOps.td:77]: All of the TfheRustBool files are done in 842
